x/gamm/keeper/pool.go

```diff
@@ -105,97 +105,110 @@ func (k Keeper) DeletePool(ctx sdk.Context, poolId uint64) error {
-// 			if err != nil {
-// 				return true
-// 			}
-// 		}
-
-// 		return false
-// 	})
-
-// 	if err != nil {
-// 		return err
-// 	}
-
-// 	for _, pool := range pools {
-// 		// sanity check
-// 		if !pool.GetTotalShares().IsZero() {
-// 			panic("pool total share should be zero after cleanup")
-// 		}
-
-// 		err = k.DeletePool(ctx, pool.GetId())
-// 		if err != nil {
-// 			return err
-// 		}
-// 	}
-
-// 	return nil
-// }
+func (k Keeper) CleanupBalancerPool(ctx sdk.Context, poolIds []uint64) (err error) {
+	// we use maps here because we can't alter the state of pool directly
+	poolAddresses := make(map[string]sdk.AccAddress)
+	totalShares := make(map[string]sdk.Int)
+	poolLiquidity := make(map[string]sdk.Coins)
+
+	for _, poolId := range poolIds {
+		pool, err := k.GetPoolAndPoke(ctx, poolId)
+		if err != nil {
+			return err
+		}
+		shareDenom := types.GetPoolShareDenom(poolId)
+
+		poolAddresses[shareDenom] = pool.GetAddress()
+		totalShares[shareDenom] = pool.GetTotalShares()
+		poolLiquidity[shareDenom] = pool.GetTotalPoolLiquidity(ctx)
+	}
+
+	// first iterate through the share holders and burn them
+	k.bankKeeper.IterateAllBalances(ctx, func(addr sdk.AccAddress, coin sdk.Coin) (stop bool) {
+		// skip to next iteration if the coin amount is zero
+		if coin.Amount.IsZero() {
+			return
+		}
+
+		// skip to next iteration if this coin is not a pool share
+		_, ok := totalShares[coin.Denom]
+		if !ok {
+			return
+		}
+
+		// track remaining total share every iteration
+		shareDenom := coin.Denom
+
+		// Burn the share tokens
+		err = k.bankKeeper.SendCoinsFromAccountToModule(ctx, addr, types.ModuleName, sdk.Coins{coin})
+		if err != nil {
+			return true
+		}
+
+		err = k.bankKeeper.BurnCoins(ctx, types.ModuleName, sdk.Coins{coin})
+		if err != nil {
+			return true
+		}
+
+		// Refund assets
+		// lpShareEquivalentTokens = (amount in pool) * (your shares) / (total shares)
+		for _, asset := range poolLiquidity[shareDenom] {
+			lpShareEquivalentTokens := asset.Amount.Mul(coin.Amount).Quo(totalShares[shareDenom])
+			if lpShareEquivalentTokens.IsZero() {
+				continue
+			}
+
+			poolAssetToReturn := sdk.NewCoin(asset.Denom, lpShareEquivalentTokens)
+			poolLiquidity[shareDenom] = poolLiquidity[shareDenom].Sub(sdk.Coins{poolAssetToReturn})
+
+			err = k.bankKeeper.SendCoins(
+				ctx, poolAddresses[shareDenom], addr, sdk.Coins{poolAssetToReturn})
+			if err != nil {
+				return true
+			}
+		}
+
+		totalShares[shareDenom] = totalShares[shareDenom].Sub(coin.Amount)
+		return false
+	})
+
+	if err != nil {
+		return err
+	}
+
+	for _, poolId := range poolIds {
+		pool, err := k.GetPoolAndPoke(ctx, poolId)
+		if err != nil {
+			return err
+		}
+		shareDenom := types.GetPoolShareDenom(poolId)
+
+		// sanity check that we have no remaining shares
+		if !poolLiquidity[shareDenom].IsZero() {
+			panic("pool total share should be zero after cleanup")
```

Comment:
consider returning an error here fore easier testing. I'm assuming the caller (upgrade handler) would panic on any error anyways

Reply:
makes sense1